src/chai/ManagedArray.inl

```diff
@@ -373,13 +359,25 @@ CHAI_INLINE
 CHAI_HOST
 void ManagedArray<T>::move(ExecutionSpace space)
 {
-  m_active_base_pointer = static_cast<T*>(m_resource_manager->move(m_active_base_pointer, m_pointer_record, space));
+  ExecutionSpace prev_space = m_pointer_record->m_last_space;
+
+  /* When moving from CPU to GPU we need to move the inner arrays before the outer array. */
+  if (prev_space == CPU) {
+    moveInnerImpl(space);
+  }
+
+  m_active_base_pointer = static_cast<T*>(m_resource_manager->move(const_cast<T_non_const*>(m_active_base_pointer), m_pointer_record, space));
   m_active_pointer = m_active_base_pointer + m_offset;
 
   if (!std::is_const<T>::value) {
     CHAI_LOG("ManagedArray", "T is non-const, registering touch of pointer" << m_active_pointer);
     m_resource_manager->registerTouch(m_pointer_record);
   }
+
+  /* When moving from GPU to CPU we need to move the inner arrays after the outer array. */
+  if (prev_space == GPU) {
```

Comment:
Sorry...Why is it the case that you need to move the outer array first? Doesn't the outer array contain invalid pointers if the inner arrays are not on the target space?

It seems oddly non-symmetric that what space is being moved from required different behavior.

Reply:
It is weird but I think that's how it has to be. 

When moving from the host to device you need to move the inner arrays first so that when you move and then memcopy the outer array-of-arrays the active pointers of the inner arrays are pointing to device memory. If you move the outer array first then the active pointers of the inner arrays on the device point to host memory and you'll segfault.

When moving from device to host you need to move the outer array first so that the active pointer of the outer array now points to host memory. Then you need to go an move each of the inner arrays (which just got memcopied over) so that their active pointers point to host memory. You can't move the inner arrays first because the active pointer of the outer array points to device memory and when you do `m_active_pointer[i] = T(m_active_pointer[i])` in `moveInnerImpl` you'll segfault.